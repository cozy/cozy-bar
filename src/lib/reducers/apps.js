import stack from 'lib/stack'
import unionWith from 'lodash.unionwith'
import EXCLUDES from 'config/excludedApps'

// constants
const DELETE_APP = 'DELETE_APP'
const RECEIVE_APP = 'RECEIVE_APP'
const RECEIVE_APP_LIST = 'RECEIVE_APP_LIST'
const RECEIVE_HOME_APP = 'RECEIVE_HOME_APP'
const FETCH_APPS = 'FETCH_APPS'
const FETCH_APPS_FAILURE = 'FETCH_APPS_FAILURE'
const SET_INFOS = 'SET_INFOS'

export const isCurrentApp = (state, app) => app.slug === state.appSlug

// selectors
export const getApps = state => {
  if (!state.apps) return []
  return state.apps
}

export const getHomeApp = state => {
  return state.homeApp
}

export const isFetchingApps = state => {
  return state ? state.isFetching : false
}

export const hasFetched = state => state.hasFetched

// actions
export const deleteApp = app => ({ type: DELETE_APP, app })
export const receiveApp = app => ({ type: RECEIVE_APP, app })
const receiveAppList = apps => ({ type: RECEIVE_APP_LIST, apps })
const receiveHomeApp = homeApp => ({ type: RECEIVE_HOME_APP, homeApp })
export const setInfos = (appName, appNamePrefix, appSlug) => ({
  type: SET_INFOS,
  appName,
  appNamePrefix,
  appSlug
})

// actions async
export const fetchApps = () => async dispatch => {
  try {
    dispatch({ type: FETCH_APPS })
    const rawAppList = await stack.get.apps()
    const apps = rawAppList
      .filter(app => !EXCLUDES.includes(app.attributes.slug))
      .map(mapApp)
    if (!rawAppList.length)
      throw new Error('No installed apps found by the bar')
    // TODO load only one time icons
    await dispatch(setHomeApp(apps))
    await dispatch(receiveAppList(apps))
  } catch (e) {
    dispatch({ type: FETCH_APPS_FAILURE })
    console.warn(e.message ? e.message : e)
  }
}

const setHomeApp = appsList => async dispatch => {
  return stack.get
    .context()
    .then(context => {
      const homeLink =
        context.data &&
        context.data.attributes &&
        context.data.attributes.default_redirection
      const slugRegexp = /^([^/]+)\/.*/
      const homeSlug =
        homeLink && homeLink.match(slugRegexp) && homeLink.match(slugRegexp)[1]
      if (!homeSlug) return appsList
      const homeApp = appsList.find(app => app.slug === homeSlug)
      return dispatch(receiveHomeApp(homeApp))
    })
    .catch(error => {
      console.warn &&
        console.warn(`Cozy-bar cannot fetch home app data: ${error.message}`)
      return appsList
    })
}

// reducers
const defaultState = {
  apps: [],
  homeApp: null,
  isFetching: true,
  appName: null,
  appNamePrefix: null,
  appSlug: null,
  hasFetched: false
}

const reducer = (state = defaultState, action) => {
  switch (action.type) {
    case FETCH_APPS:
      return { ...state, isFetching: true }
    case FETCH_APPS_FAILURE:
      return { ...state, isFetching: false }
    case RECEIVE_APP:
      return {
        ...state,
        apps: unionWith(
          state.apps,
          [mapApp(action.app)],
          (appA, appB) => appA.slug === appB.slug
        )
      }
    case RECEIVE_APP_LIST: {
      const appsList = action.apps.map(app => ({
        ...app,
        isCurrentApp: isCurrentApp(state, app)
      }))
      return { ...state, isFetching: false, hasFetched: true, apps: appsList }
    }
    case RECEIVE_HOME_APP: {
      const homeApp = action.homeApp
      return isCurrentApp(state, homeApp)
        ? {
            ...state,
            homeApp: { ...homeApp, isCurrentApp: true }
          }
        : { ...state, homeApp }
    }
    case DELETE_APP:
      return {
        ...state,
        apps: state.apps.filter(app => app.slug !== action.app.slug)
      }
    case SET_INFOS:
      return {
        ...state,
        appName: action.appName,
        appNamePrefix: action.appNamePrefix,
        appSlug: action.appSlug
      }
    default:
      return state
  }
}

export default reducer

// helpers
const camelCasify = object =>
  !!object &&
  Object.keys(object).reduce((acc, key) => {
    const camelCaseKey = key
      .split('_')
      .map((segment, index) =>
        index ? segment.charAt(0).toUpperCase() + segment.slice(1) : segment
      )
      .join('')
    acc[camelCaseKey] = object[key]
    return acc
  }, {})

const mapApp = app => ({
  ...app,
  ...camelCasify(app.attributes),
  href: app.links && app.links.related
})
