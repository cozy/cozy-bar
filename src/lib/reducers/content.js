const SET_CONTENT = 'SET_CONTENT'
const UNSET_CONTENT = 'UNSET_CONTENT'

const getLastItemInMap = map => Array.from(map)[map.size - 1]

// action creator
export const setContent = (location, content, id) => ({
  type: SET_CONTENT,
  location,
  content,
  id
})
export const unsetContent = (location, id) => ({
  type: UNSET_CONTENT,
  location,
  id
})

// reducer
export const getDefaultState = () => ({
  left: new Map(),
  center: new Map(),
  right: new Map(),
  search: new Map()
})

export const reducer = (state = getDefaultState(), action) => {
  if (!action.location || typeof action.id === undefined) return state
  switch (action.type) {
    case SET_CONTENT: {
      const currentState = state[action.location]
      currentState.set(action.id, action.content)
      return { ...state, [action.location]: currentState }
    }
    case UNSET_CONTENT: {
      const currentState = state[action.location]
      if (!currentState.get(action.id)) {
        return state
      }
      currentState.delete(action.id)
      return { ...state, [action.location]: currentState }
    }
    default:
      return state
  }
}

// selectors
export const getContent = (state, location) =>
  getLastItemInMap(state[location]) && getLastItemInMap(state[location])[1]
