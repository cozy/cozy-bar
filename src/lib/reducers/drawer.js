// constants
const OPEN_DRAWER = 'OPEN_DRAWER'
const CLOSE_DRAWER = 'CLOSE_DRAWER'

// actions
export const openDrawer = () => ({ type: OPEN_DRAWER })
export const closeDrawer = () => ({ type: CLOSE_DRAWER })

// selectors
export const isDrawerOpen = state => state.isOpen

// reducers
const defaultState = {
  isOpen: false
}

const reducer = (state = defaultState, action) => {
  switch (action.type) {
    case OPEN_DRAWER:
      return { isOpen: true }

    case CLOSE_DRAWER:
      return { isOpen: false }

    default:
      return state
  }
}

export default reducer
