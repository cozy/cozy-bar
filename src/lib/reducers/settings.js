import stack from 'lib/stack'

const FETCH_SETTINGS = 'FETCH_SETTINGS'
const FETCH_SETTINGS_BUSY = 'FETCH_SETTINGS_BUSY'
const FETCH_SETTINGS_SUCCESS = 'FETCH_SETTINGS_SUCCESS'
const RECEIVE_NO_CONTEXT = 'RECEIVE_NO_CONTEXT'
const RECEIVE_STORAGE = 'RECEIVE_STORAGE'
const RECEIVE_SETTINGS_URL = 'RECEIVE_SETTINGS_URL'

export const LOG_OUT = 'LOG_OUT'

const BUSY_DELAY = 450

// selectors
export const getStorageData = state => {
  return state.storageData
}
export const getSettingsAppURL = state => {
  return state.settingsAppURL
}
export const isSettingsBusy = state => {
  return state.isBusy
}
export const isFetchingSettings = state => {
  return state.isFetching
}

// actions
const fetchStorageData = () => async dispatch => {
  try {
    const storageData = await stack.get.storageData()
    return dispatch({ type: RECEIVE_STORAGE, storageData })
  } catch (e) {
    // eslint-disable-next-line no-console
    console.warn && console.warn('Cannot get Cozy storage informations')
    return null
  }
}

const fetchSettingsAppURL = () => async (dispatch, getState) => {
  // If the `settings` app is available, it will used to add the links 'Profile' and 'Connected Devices'
  if (getState().cozyBar.settings.settingsAppURL) {
    return dispatch({
      type: RECEIVE_SETTINGS_URL,
      settingsAppURL: getState().cozyBar.settings.settingsAppURL
    })
  }
  try {
    const settingsAppURL = await stack.get.settingsAppURL()
    return dispatch({ type: RECEIVE_SETTINGS_URL, settingsAppURL })
  } catch (e) {
    // eslint-disable-next-line no-console
    console.warn('Settings app is unavailable, settings links are disabled')
    return null
  }
}

export const fetchSettingsData = (displayBusy = true) => async dispatch => {
  dispatch({ type: FETCH_SETTINGS })
  // put the busy status after BUSY_DELAY secs
  const busySpinner = setTimeout(() => {
    // we do not display the busy status in the drawer
    if (displayBusy) dispatch({ type: FETCH_SETTINGS_BUSY })
  }, BUSY_DELAY)
  await dispatch(fetchStorageData())
  await dispatch(fetchSettingsAppURL())
  clearTimeout(busySpinner)
  dispatch({ type: FETCH_SETTINGS_SUCCESS })
}

export const logOut = () => async dispatch => {
  dispatch({ type: LOG_OUT })
  try {
    await stack.logout()
  } catch (e) {
    // eslint-disable-next-line no-console
    console.warn('Error while logging out in the cozy-bar', e)
  }
}

// reducers
const defaultState = {
  contextNotExist: false,
  isFetching: false,
  isBusy: false,
  settingsAppURL: null,
  storageData: null
}

const reducer = (state = defaultState, action) => {
  switch (action.type) {
    case FETCH_SETTINGS:
      return { ...state, isFetching: true }
    case FETCH_SETTINGS_BUSY:
      return { ...state, isBusy: true }
    case FETCH_SETTINGS_SUCCESS:
      return { ...state, isFetching: false, isBusy: false }
    case RECEIVE_NO_CONTEXT:
      return { ...state, contextNotExist: true }
    case RECEIVE_STORAGE:
      return { ...state, storageData: action.storageData }
    case RECEIVE_SETTINGS_URL:
      return { ...state, settingsAppURL: action.settingsAppURL }
    case LOG_OUT:
      return defaultState
    default:
      return state
  }
}

export default reducer
