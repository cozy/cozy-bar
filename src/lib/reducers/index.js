import { combineReducers } from 'redux'
import * as content from 'lib/reducers/content'
import * as locale from 'lib/reducers/locale'
import * as theme from 'lib/reducers/theme'
import appsReducer, * as apps from 'lib/reducers/apps'
import settingsReducer, * as settings from 'lib/reducers/settings'
import contextReducer, * as context from 'lib/reducers/context'

const proxy = (attr, method) => {
  return (state, ...args) => {
    return method(state[attr], ...args)
  }
}

const setContent = content.setContent
const unsetContent = content.unsetContent
const setLocale = locale.setLocale
const setTheme = theme.setTheme
const fetchApps = apps.fetchApps
const setInfos = apps.setInfos
const fetchSettingsData = settings.fetchSettingsData
const logOut = settings.logOut
const fetchContext = context.fetchContext
export {
  setContent,
  unsetContent,
  setLocale,
  setTheme,
  fetchApps,
  setInfos,
  fetchSettingsData,
  logOut,
  fetchContext
}

export const getContent = proxy('content', content.getContent)
export const getLocale = proxy('locale', locale.getLocale)
export const getTheme = proxy('theme', theme.getTheme)
export const getApps = proxy('apps', apps.getApps)
export const getHomeApp = proxy('apps', apps.getHomeApp)
export const isFetchingApps = proxy('apps', apps.isFetchingApps)
export const isCurrentApp = proxy('apps', apps.isCurrentApp)
export const hasFetched = proxy('apps', apps.hasFetched)
export const getStorageData = proxy('settings', settings.getStorageData)
export const getSettingsAppURL = proxy('settings', settings.getSettingsAppURL)
export const isSettingsBusy = proxy('settings', settings.isSettingsBusy)
export const isFetchingSettings = proxy('settings', settings.isFetchingSettings)
export const getHelpLink = proxy('context', context.getHelpLink)
export const getClaudyActions = proxy('context', context.getClaudyActions)
export const shouldEnableClaudy = proxy('context', context.shouldEnableClaudy)

// realtime handlers
export const onRealtimeCreate = apps.receiveApp
export const onRealtimeDelete = apps.deleteApp

export const reducers = {
  content: content.reducer,
  locale: locale.reducer,
  theme: theme.reducer,
  apps: appsReducer,
  settings: settingsReducer,
  context: contextReducer
}

export default combineReducers(reducers)
