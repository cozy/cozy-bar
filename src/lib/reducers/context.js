import stack from 'lib/stack'
import CLAUDY_ACTIONS from 'config/claudyActions.json'
import { LOG_OUT } from 'lib/reducers/settings'

const FETCH_CONTEXT = 'FETCH_CONTEXT'
const FETCH_CONTEXT_SUCCESS = 'FETCH_CONTEXT_SUCCESS'
const RECEIVE_NO_CONTEXT = 'RECEIVE_NO_CONTEXT'

// selectors
export const getHelpLink = state => {
  return state.helpLink
}

export const getClaudyActions = state => {
  return state.claudyActions
}
export const shouldEnableClaudy = state => {
  const claudyActions = getClaudyActions(state)
  return !!claudyActions && !!claudyActions.length
}

// actions
export const fetchContext = () => async (dispatch, getState) => {
  dispatch({ type: FETCH_CONTEXT })
  if (getState().context.contextNotExist) {
    return dispatch({ type: FETCH_CONTEXT_SUCCESS, context: null })
  }
  try {
    const context = await stack.get.context()
    return dispatch({ type: FETCH_CONTEXT_SUCCESS, context })
  } catch (error) {
    if (error.status && error.status === 404) {
      dispatch({ type: RECEIVE_NO_CONTEXT })
    }
    // eslint-disable-next-line no-console
    console.warn('Cannot get Cozy context')
    return null
  }
}

// reducers
const defaultState = {
  claudyActions: [],
  contextNotExist: false,
  helpLink: null,
  isFetching: false
}

const reducer = (state = defaultState, action) => {
  switch (action.type) {
    case FETCH_CONTEXT:
      return { ...state, isFetching: true }
    case FETCH_CONTEXT_SUCCESS: {
      const attr =
        action.context && action.context.data && action.context.data.attributes
      const contextActions = (attr && attr['claudy_actions']) || []
      // get an arrays of action
      const claudyActions = contextActions
        .map(slug => {
          if (CLAUDY_ACTIONS.hasOwnProperty(slug)) {
            // adding also the action slug
            return Object.assign({}, CLAUDY_ACTIONS[slug], { slug })
          }
        })
        .filter(action => action)
      return {
        ...state,
        helpLink: (attr && attr['help_link']) || null,
        claudyActions,
        isFetching: false,
        contextNotExist: false
      }
    }
    case RECEIVE_NO_CONTEXT:
      return { ...state, contextNotExist: true }
    case LOG_OUT:
      return defaultState
    default:
      return state
  }
}

export default reducer
