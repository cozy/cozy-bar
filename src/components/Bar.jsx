/* global __TARGET__ */
/* global __PIWIK_TRACKER_URL__  __PIWIK_SITEID__ __PIWIK_DIMENSION_ID_APP__ */

import React, { Component } from 'react'
import { translate } from '../lib/I18n'
import { shouldEnableTracking, getTracker, configureTracker } from 'cozy-ui/react/helpers/tracker'

import Drawer from './Drawer'
import Nav from './Nav'
import Claudy from './Claudy'

class Bar extends Component {
  constructor (props, context) {
    super(props)
    this.store = context.store
    this.state = {
      enableClaudy: null, // no claudy by default
      fireClaudy: false, // true to fire claudy (used by the drawer)
      claudyOpened: false,
      drawerVisible: false,
      usageTracker: null
    }
    this.toggleDrawer = this.toggleDrawer.bind(this)
  }

  async componentWillMount () {
    const enableClaudy = await this.store.shouldEnableClaudy()
    this.setState({ enableClaudy })
  }

  componentDidMount () {
    // if tracking enabled, init the piwik tracker
    if (shouldEnableTracking()) {
      const trackerInstance = getTracker(__PIWIK_TRACKER_URL__, __PIWIK_SITEID__, false, false)
      configureTracker({
        appDimensionId: __PIWIK_DIMENSION_ID_APP__,
        app: 'Cozy Bar',
        heartbeat: 0
      })
      this.setState({ usageTracker: trackerInstance })
    }
  }

  toggleDrawer () {
    // don't allow to toggle the drawer if claudy opened or is opening
    if (this.state.claudyOpened || this.state.fireClaudy) return
    const drawerVisible = !this.state.drawerVisible
    // don't wait for transitionend if displaying
    if (drawerVisible) this.props.onDrawer(drawerVisible)
    this.setState({ drawerVisible })
  }

  toggleClaudy (isFromDrawer = false) {
    if (!this.state.enableClaudy) return
    const { usageTracker, claudyOpened } = this.state
    if (isFromDrawer && !claudyOpened) { // if opened from drawer
      // reset to toggle via the Claudy component
      return this.setState({fireClaudy: true})
    }
    if (this.state.fireClaudy) this.setState({fireClaudy: false})
    if (usageTracker) {
      usageTracker.push([
        'trackEvent',
        'Claudy',
        claudyOpened ? 'close' : 'open',
        'claudy'
      ])
    }
    this.setState({ claudyOpened: !claudyOpened })
  }

  render () {
    const { t, lang, appName,
      appEditor, iconPath, replaceTitleOnMobile,
      onDrawer, isPublic } = this.props
    const { usageTracker, claudyOpened,
      enableClaudy, drawerVisible, fireClaudy } = this.state
    return (
      <div class='coz-bar-container'>
        <h1 lang={lang} class={`coz-bar-title ${replaceTitleOnMobile ? 'coz-bar-hide-sm' : ''}`}>
          <img class='coz-bar-hide-sm' src={iconPath} width='32' />
          {appEditor && <span class='coz-bar-hide-sm'>{appEditor} </span>}
          <strong>{appName}</strong>
          <sup class='coz-bar-hide-sm coz-bar-beta-status'>{t('beta')}</sup>
        </h1>
        {__TARGET__ !== 'mobile' && !isPublic &&
          <div class='coz-bar-flex-container'>
            <button class='coz-bar-burger' onClick={this.toggleDrawer} data-icon='icon-hamburger'>
              <span class='coz-bar-hidden'>{t('drawer')}</span>
            </button>
            <Drawer visible={drawerVisible} onClose={this.toggleDrawer} onClaudy={(enableClaudy && (() => this.toggleClaudy(true))) || false} isClaudyLoading={fireClaudy} drawerListener={() => onDrawer(this.state.drawerVisible)} />
            <Nav />
            {enableClaudy &&
              <Claudy
                usageTracker={usageTracker}
                fireClaudy={fireClaudy}
                onToggle={() => this.toggleClaudy(false)}
                opened={claudyOpened}
              />
            }
          </div>
        }
      </div>
    )
  }
}

export default translate()(Bar)
