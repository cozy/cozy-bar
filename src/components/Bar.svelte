<button class='coz-bar-burger' on:click='toggleDrawer()' data-icon='icon-hamburger'>
  {{t('menu')}}
</button>

<h1 class='coz-bar-title'>
  <img class='coz-bar-hide-sm' src='{{iconPath}}' width='32' />
  <span class='coz-bar-hide-sm'>cozy </span><strong>{{appName}}</strong>
</h1>

<hr class='coz-sep-flex' />

<Navigation sections='{{sections}}' on:open='onPopOpen(event.panel)' />

<Drawer ref:drawer content='{{config.apps.length?config.apps[0]:false}}' footer='{{sections[1].items}}' />

<script>
  import { t } from '../lib/i18n'
  import stack from '../lib/stack'

  import Navigation from './Navigation'
  import Drawer from './Drawer'

  import MENU_CONFIG from '../config/menu'

  async function updateAppsItems () {
    try {
      let apps = await stack.get.apps()
      apps = [await apps.map(app => {
        return {
          label: app.attributes.name,
          external: true,
          l10n: false,
          href: `https://${app.attributes.slug}.cozy.local/`,
          icon: `https://${app.attributes.slug}.cozy.local/${app.attributes.icon}`
        }
      })]
      this.set({ config: Object.assign({}, this.get('config'), {apps}) })
    } catch (e) {
      console.error(e.message)
    }
  }

  async function updateDiskUsage () {
      let currentDiskUsage
    try {
      let data = await stack.get.diskUsage()
      currentDiskUsage = parseInt(data.attributes.used)
    } catch (e) {
      console.error(e.message)
      currentDiskUsage = { error: e.name }
    }
    // copy settings section to update storage item
    const newSettings = this.get('config').settings.slice()
    newSettings.forEach(section => {
      let storageItem = section.find(item => item.label === 'storage')
      if (storageItem) {
        storageItem.currentDiskUsage = currentDiskUsage
      }
    })
    this.set({ config: Object.assign({}, this.get('config'), {settings: newSettings}) })
  }

  export default {
    data() {
      return {
        config: MENU_CONFIG
      }
    },

    computed: {
      sections: config => {
        return [{
          label: 'apps',
          icon: 'icon-cube',
          items: config.apps
        }, {
          label: 'settings',
          icon: 'icon-cog',
          items: config.settings
        }]
      }
    },

    components: {
      Navigation,
      Drawer
    },

    helpers: { t },

    methods: {
      toggleDrawer() {
        this.refs.drawer.set({folded: false})
        updateAppsItems.call(this)
      },
      onPopOpen (panel) {
        if (panel === 'apps') {
          updateAppsItems.call(this)
        } else if (panel === 'settings') {
          updateDiskUsage.call(this)
        }
      }
    }
  }
</script>
