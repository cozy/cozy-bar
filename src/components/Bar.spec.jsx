import React from 'react'
import { shallow } from 'enzyme'
import { isMobileApp } from 'cozy-device-helper'
import toJson from 'enzyme-to-json'
import reducers from 'lib/reducers'
import CozyClient from 'cozy-client'

import { Bar, mapStateToProps, mapDispatchToProps } from './Bar'

jest.mock('cozy-device-helper', () => ({
  ...require.requireActual('cozy-device-helper'),
  isMobileApp: jest.fn()
}))

describe('Bar', () => {
  let props, client
  beforeEach(() => {
    jest.resetAllMocks()
    jest.spyOn(Bar.prototype, 'fetchApps')
    isMobileApp.mockReturnValue(false)
    client = new CozyClient({})
    props = {
      fetchContext: jest.fn().mockResolvedValue({}),
      fetchApps: jest.fn().mockResolvedValue([]),
      fetchSettingsData: jest.fn().mockResolvedValue({}),
      theme: 'default',
      themeOverrides: {},
      t: x => x,
      cozyClient: client
    }
  })

  afterEach(() => {
    Bar.prototype.fetchApps.mockRestore()
  })

  it('should fetch data when mounted', () => {
    shallow(<Bar {...props} />)
    expect(props.fetchContext).toHaveBeenCalled()
    expect(props.fetchApps).toHaveBeenCalled()
    expect(props.fetchSettingsData).toHaveBeenCalled()
    expect(Bar.prototype.fetchApps).toHaveBeenCalled()
  })

  it('should not fetch data if public', () => {
    shallow(<Bar {...props} isPublic={true} />)
    expect(props.fetchContext).not.toHaveBeenCalled()
    expect(props.fetchApps).not.toHaveBeenCalled()
    expect(props.fetchSettingsData).not.toHaveBeenCalled()
    expect(Bar.prototype.fetchApps).not.toHaveBeenCalled()
  })

  it('should not fetch apps if hasFetchedApps is true', () => {
    const barWrapper = shallow(<Bar {...props} hasFetchedApps={true} />)
    jest.resetAllMocks()
    barWrapper.setState({ drawerVisible: true }) // to call componentdidUpdate
    expect(props.fetchApps).not.toHaveBeenCalled()
    expect(Bar.prototype.fetchApps).not.toHaveBeenCalled()
  })

  it('should re-fetch apps if apps are not fetched and drawer opens', () => {
    const barWrapper = shallow(<Bar {...props} />)
    expect(Bar.prototype.fetchApps).toHaveBeenCalledTimes(1)
    expect(props.fetchApps).toHaveBeenCalledTimes(1)
    barWrapper.setState({ drawerVisible: true })
    expect(props.fetchApps).toHaveBeenCalledTimes(2)
    expect(Bar.prototype.fetchApps).toHaveBeenCalledTimes(2)
    barWrapper.setState({ drawerVisible: true, usageTracker: {} })
    expect(props.fetchApps).toHaveBeenCalledTimes(2)
    expect(Bar.prototype.fetchApps).toHaveBeenCalledTimes(2)
  })

  it('should change theme', () => {
    const barWrapper = shallow(<Bar {...props} theme="primary" />)
    expect(toJson(barWrapper)).toMatchSnapshot()
  })

  it('should change allow theme overrides', () => {
    const barWrapper = shallow(
      <Bar
        {...props}
        theme="primary"
        themeOverrides={{ primaryColor: 'red' }}
      />
    )
    expect(toJson(barWrapper)).toMatchSnapshot()
  })

  it('should display the Searchbar', () => {
    const barWrapper = shallow(<Bar {...props} isDrive isPublic={false} />)
    expect(toJson(barWrapper)).toMatchSnapshot()
  })

  it('should not display searchbar if we are on mobile', () => {
    isMobileApp.mockReturnValue(true)
    const barWrapper = shallow(<Bar {...props} isDrive isPublic={false} />)
    expect(toJson(barWrapper)).toMatchSnapshot()
  })

  it('should not display searchbar if we are not on Cozy Drive', () => {
    isMobileApp.mockReturnValue(true)
    const barWrapper = shallow(
      <Bar {...props} isDrive={false} isPublic={false} />
    )
    expect(toJson(barWrapper)).toMatchSnapshot()
  })

  it('should not display searchbar if we are not on  a public page', () => {
    isMobileApp.mockReturnValue(true)
    const barWrapper = shallow(<Bar {...props} isDrive isPublic={true} />)
    expect(toJson(barWrapper)).toMatchSnapshot()
  })

  it('should have correct state props provided by the store with the initial state', () => {
    const initialState = reducers(undefined, {})
    expect(mapStateToProps(initialState)).toMatchSnapshot()
  })

  it('should have correct dispatch props provided by the store', () => {
    expect(mapDispatchToProps(jest.fn())).toMatchSnapshot()
  })

  it('should call re-fetch data when token is refreshed', () => {
    shallow(
      <Bar {...props} isDrive={false} isPublic={false} />
    )
    client.emit('tokenRefreshed')
    expect(props.fetchContext).toHaveBeenCalledTimes(2)
    expect(props.fetchApps).toHaveBeenCalledTimes(2)
    expect(props.fetchSettingsData).toHaveBeenCalledTimes(2)
    expect(Bar.prototype.fetchApps).toHaveBeenCalledTimes(2)
  })
})
