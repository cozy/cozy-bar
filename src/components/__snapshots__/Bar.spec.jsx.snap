// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bar should change allow theme overrides 1`] = `
<div
  className="coz-bar-wrapper coz-theme-primary"
  style={
    Object {
      "--cozBarThemePrimaryColor": "red",
    }
  }
>
  <div
    id="cozy-bar-modal-dom-place"
  />
  <div
    className="coz-bar-container"
  >
    <button
      className="coz-bar-btn coz-bar-burger"
      data-tutorial="apps-mobile"
      onClick={[Function]}
      type="button"
    >
      <Icon
        color="currentColor"
        height={16}
        icon={[Function]}
        spin={false}
        width={16}
      />
      <span
        className="coz-bar-hidden"
      >
        drawer
      </span>
    </button>
    <Apps />
    <div
      className="u-flex-grow"
    />
    <ForwardRef />
    <Connect(Drawer)
      drawerListener={[Function]}
      isClaudyLoading={false}
      onClaudy={false}
      onClose={[Function]}
      visible={false}
    />
  </div>
</div>
`;

exports[`Bar should change theme 1`] = `
<div
  className="coz-bar-wrapper coz-theme-primary"
  style={Object {}}
>
  <div
    id="cozy-bar-modal-dom-place"
  />
  <div
    className="coz-bar-container"
  >
    <button
      className="coz-bar-btn coz-bar-burger"
      data-tutorial="apps-mobile"
      onClick={[Function]}
      type="button"
    >
      <Icon
        color="currentColor"
        height={16}
        icon={[Function]}
        spin={false}
        width={16}
      />
      <span
        className="coz-bar-hidden"
      >
        drawer
      </span>
    </button>
    <Apps />
    <div
      className="u-flex-grow"
    />
    <ForwardRef />
    <Connect(Drawer)
      drawerListener={[Function]}
      isClaudyLoading={false}
      onClaudy={false}
      onClose={[Function]}
      visible={false}
    />
  </div>
</div>
`;

exports[`Bar should have correct dispatch props provided by the store 1`] = `
Object {
  "fetchApps": [Function],
  "fetchContext": [Function],
  "fetchSettingsData": [Function],
}
`;

exports[`Bar should have correct state props provided by the store with the initial state 1`] = `
Object {
  "barCenter": undefined,
  "barLeft": undefined,
  "barRight": undefined,
  "barSearch": undefined,
  "claudyEnabled": false,
  "hasFetchedApps": false,
  "isDrive": false,
  "theme": "default",
  "themeOverrides": Object {},
  "webviewContext": undefined,
}
`;
